### PR TITLE
Connecting EventHubBufferedProducerClient to internal classes.

### PR DESCRIPTION
**Pull request by conniey**
# Description

* Connecting EventHubBufferedProducerClient and EventHubBufferedProducerAsyncClient to its internal classes
* Fixing log messages to use fluent logger.
* Make features not released for beta package-private.

Related #24156

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventHubBufferedPartitionProducer.java [hunk 1]
```diff
@@ -15,87 +16,129 @@
...
-            });
+            } catch (InterruptedException e) {
+                // Unsure whether this is recoverable by trying again? Maybe, since this could be scheduled on
+                // another thread.
+                sink.error(new TimeoutException("Unable to acquire flush semaphore due to interrupted exception."));
+                return;
+            }
+
+            try {
+                if (isClosed.get()) {
```
Comment (ZejiaJiang) [inline sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventHubBufferedPartitionProducer.java:112]: nit: why we don't check isClosed before isflushing?

File: sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventHubBufferedProducerClientBuilder.java [hunk 2]
```diff
@@ -353,14 +367,40 @@ public EventHubBufferedProducerClientBuilder transportType(AmqpTransportType tra
...
+     *     #onSendBatchFailed(Consumer)}, or {@link #maxWaitTime(Duration)} are null.
+     * @throws IllegalArgumentException if {@link #maxConcurrentSends(int)}, {@link
+     *     #maxConcurrentSendsPerPartition(int)}, or {@link #maxEventBufferLengthPerPartition(int)} are less than 1.
      */
     public EventHubBufferedProducerAsyncClient buildAsyncClient() {
-        return new EventHubBufferedProducerAsyncClient(builder, clientOptions);
+
+        if (Objects.isNull(clientOptions.getSendSucceededContext())) {
+            throw LOGGER.logExceptionAsError(new NullPointerException("'onSendBatchSucceeded' cannot be null."));
+        } else if (Objects.isNull(clientOptions.getSendFailedContext())) {
```
Comment (liukun-msft) [inline sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventHubBufferedProducerClientBuilder.java:384]: remove all the `else` keywords? It looks like they are `and` relationship not `or-else` relationship?

File: sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventHubBufferedPartitionProducer.java [hunk 3]
```diff
@@ -15,87 +16,129 @@
...
-    private final Retry retryWhenPolicy = Retry.from(signal -> {
-        if (isClosed.get()) {
-            return Mono.empty();
-        } else {
-            return Mono.just(true);
-        }
-    });
     private final CreateBatchOptions createBatchOptions;
     private final Queue<EventData> eventQueue;
+    private final AtomicBoolean flush = new AtomicBoolean(false);
```
Comment (liukun-msft) [inline sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventHubBufferedPartitionProducer.java:57]: we have a function `flush()`, change to `isFlushed` for boolean name?

**Review by liukun-msft — COMMENTED** [review]
Looks great! Add few comments from my end.

File: sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventHubBufferedPartitionProducer.java (hunk 3 above)
Comment (conniey) [reply sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventHubBufferedPartitionProducer.java:57]: When I was naming this variable, I kept thinking of the bathroom. 😄 But you're right, to align, I'll rename this.

File: sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventHubBufferedProducerClient.java [hunk 4]
```diff
@@ -41,5 +46,191 @@
...
+
+    /**
+     * Attempts to publish all events in the buffer immediately.  This may result in multiple batches being published,
+     * the outcome of each of which will be individually reported by the {@link EventHubBufferedProducerClientBuilder#onSendBatchFailed(Consumer)}
+     * and {@link EventHubBufferedProducerClientBuilder#onSendBatchSucceeded(Consumer)} handlers.
+     *
+     * Upon completion of this method, the buffer will be empty.
+     */
+    public void flush() {
+        client.flush();
```
[the remark this answers is not in the document]
Comment (conniey) [reply sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventHubBufferedProducerClient.java:226]: Good catch!

[the remark this answers is not in the document]
Comment (conniey) [reply sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventHubBufferedProducerClient.java:226]: Oop. it actually returns void. :)

File: sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventHubBufferedProducerClientBuilder.java (hunk 2 above)
Comment (conniey) [reply sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventHubBufferedProducerClientBuilder.java:384]: Sure. I was trying to save white space. :) But fixed.

**Review by anushkasingh16 — APPROVED** [review]
Looks good :D

**Pull request merged**